### ISSUE TITLE
Should `throw throw x` be disallowed?

### ISSUE THREAD

**Issue by rbuckton** [rbuckton/proposal-throw-expressions#9]
In #7 there was a question as to whether `throw throw x` should be disallowed.

**Comment by rbuckton** [comment]
I believe that it should be allowed, as there are no issues with the runtime semantics and the syntax behaves similarly to `void void x`, which is allowed today. Disallowing it would be counterintuitive in relation to other unary expressions. Just disallowing `throw throw x` wouldn't prevent `throw (throw x)` or `throw (0, throw x)` or `throw void throw x`, etc. Trying to disallow this in the spec would be a wild goose chase and would overcomplicate the specification.

It seems that disallowing `throw throw x` would be better left as a lint rule.

**Comment by MadaraUchiha** [comment]
I tend to agree. In userland terms, it's similar to passing the result of a `void` returning function to another void returning function (a-la `console.log(console.log('foo'))`), which is allowed today.

Seeing how `throw` currently accepts any expression and `throw x` is proposed to be a valid expression, I agree that the syntax should be valid.

Another interesting question though is what's the behavior in this case. The practical, expected behavior in my opinion is that the innermost operand to `throw throw throw .... throw x` is thrown in reality, and that the extra `throw`s are no-ops.

**Comment by ljharb** [comment]
They'd be noops at runtime, but a debugger that pauses on exceptions might pause on each one.

**Comment by MadaraUchiha** [comment]
It's worth noting that in C# (which already has support for `throw` expressions), it is *not* valid syntax to `throw throw x`, *however*, I suspected that it's because by default `throw` is still a statement and not an expression, so fiddling with it a bit I came up with 

```Csharp
    string foo = null;
    var myException = new Exception();
    var bar = foo ?? throw foo ?? throw myException; // to trick it into throw expression mode
```

And that *still* does not compile, although not because of syntax error. It's because the result of `throw myException` does not inherit from `System.Exception` and thus cannot be throw'd.

So in the context of throw expressions, they are *technically* valid syntax in C#, just that it's pointless because of the type system having your back.

**Comment by MadaraUchiha** [comment]
I'll correct that. The reason that my example doesn't work in C# is because the resulting type of `foo ?? throw` in this case is `string | Never`, which resolves to `string`.

```C#
    Exception foo = null;
    var myException = new Exception();
    var bar = foo ?? throw foo ?? throw myException; // to trick it into throw expression mode
```

This compiles *and* actually works as I proposed in runtime (that is, the outer throw is no-op).

**Comment by MadaraUchiha** [comment]
Why would a debugger step into all of the `throw`s? Only the innermost one is actually reached in the execution context.

Consider
```js
throw hello();

function hello() {
  throw new Error();
}
```

would you expect the debugger to step through both in this case?

**Comment by ljharb** [comment]
If the debugger paused on each throw - not each catch - yes, i would.

**Comment by MadaraUchiha** [comment]
But the outer throws are never reached, order of execution inside-out in this case, I wouldn't expect `debugger;` to have been hit in this scenario, because the parameter is evaluated first, and then `throw` ends the execution context:
```js
function foo(param) {
  debugger;
  console.log(param);
}

foo(throw new Error());
```

As a side note: Enabling pause on exceptions with the C# example I've given above does not pause on the outer throw.

**Comment by MadaraUchiha** [comment]
Also, the example I gave in 
https://github.com/tc39/proposal-throw-expressions/issues/9#issuecomment-416549366 is valid JavaScript, the debugger doesn't pause on the outer `throw` (as is expected, in my opinion).

**Comment by ljharb** [comment]
Fair point, given that the outer throw is bypassed.

**Comment by neyb** [comment]
I don't see why `throw throw x` should be disallowed, while `throw (()=>{throw x})()` is allowed...

I think throw throw x is strange but totally ok if throw is an expression. I also think we have a consensus on this.

Can we close this (non-)issue ?

**Comment by ljharb** [comment]
There is simply no relationship between those two examples, and the second doesn’t justify the first.

The committee has indeed discussed this, and it seems it will be allowed, but this is very much not a “non-issue”, and bringing up things like this are a critical part of the proposal process.

**Comment by MadaraUchiha** [comment]
> There is simply no relationship between those two examples, and the second doesn’t justify the first.
> 
> The committee has indeed discussed this, and it seems it will be allowed, but this is very much not a “non-issue”, and bringing up things like this are a critical part of the proposal process.

I don't understand why. Why would the two examples be any different? For that matter, I would imagine `throw throw x` to be directly transpiled into `(() => { throw (() => { throw x; }); })();` by a babel plugin.

Can you provide a case where you would expect anything different?

**Comment by ljharb** [comment]
Because `throw throw x` is always nonsense, while throwing a function call is only nonsense if it happens to throw unconditionally (and swapping an identifier for the function must behave the same).

`throw throw x` can always be safely transpiled into `throw x`, in fact, because more than one throw keyword is purely redundant.

**Comment by MadaraUchiha** [comment]
How do you know ahead of time that it's nonsense? You're not expecting to know ahead of time that throwing a function call would be pointless, why do you expect to know ahead of time that throwing an expression (that happens to be another `throw`) is pointless?

**Comment by ljharb** [comment]
@MadaraUchiha because it's syntactic, and so the parser can immediately know that `throw throw throw throw throw throw throw throw throw throw x` and `throw x` are literally identical.

**Comment by MadaraUchiha** [comment]
> @MadaraUchiha because it's syntactic, and so the parser can immediately know that `throw throw throw throw throw throw throw throw throw throw x` and `throw x` are literally identical.

That's an implementation detail, a powerful enough analyzer would be able to eliminate `throw (() => { throw x; })` as well.

To me, any expression is a conceptual function call, which returns whatever the expression returns.

And in any case, there are very few (if any, I can't think of any) cases in the language where a literal is treated differently from an expression that "returns" it, and definitely not where a generic expression is allowed but a literal does not. 

I'm not sure the exception (heh) in the parser/engine is worth it, just to throw *anyway* since how else you would prevent `throw throw x` from being used?

**Comment by Jack-Works** [comment]
Allowing them make no harm to the language and make spec, parsers, ... simpler. I don't see the point why it should be banned.

**Comment by ljharb** [comment]
The committee feels the same. That doesn’t mean this is a “non-issue” or that it shouldn’t have been brought up, is all.

**Comment by rattrayalex** [comment]
Should this issue be closed?

**Comment by MadaraUchiha** [comment]
> 
> 
> Should this issue be closed?

Is there anything further to discuss on this matter?

@ljharb @Jack-Works

**Comment by alban-semana** [comment]
### tl;dr
no, it's legal, but tools should provide a warning

### simple and quick explanation
`x` is an expression and `throw x` is allowed.
the same applies; to `throw x` is an expression so `throw throw x` is allowed. 

### an example for the explanation :  
Imagine running `throw x()` (which is allowed atm):  

When running the code , 2 things are made : **evaluating `x()`** and **throwing this evaluation result**:

when evaluating `x()`, _it could fail_ and throw an error (imagine `x()` implementation being `x = () => {throw y}`), then your second step (throwing the result) will not happen.

 the conclusion is: if you allow `throw x()`, you should allow `throw throw x()`

### second example even simpler
open your console and write `throw notExistingVar` : your code is exactly the same as `notExistingVar` because your code fails at runtime **before** running the `throw` because the right operand is in error, still writing `throw notExistingVar` is allowed

@rbuckton : Is it clear enough to close this issue ?

**Comment by neyb** [comment]
### tl;dr
no, it should be legal, but tools should provide a warning

### simple and quick explanation
`x` is an expression and `throw x` is allowed.
the same applies; to `throw x` is an expression so `throw throw x` is allowed. 

### an example for the explanation :  
Imagine running `throw x()` (which is allowed atm):  

When running the code , 2 things are made : **evaluating `x()`** and **throwing this evaluation result**:

when evaluating `x()`, _it could fail_ and throw an error (imagine `x()` implementation being `x = () => {throw y}`), then your second step (throwing the result) will not happen.

 the conclusion is: if you allow `throw x()`, you should allow `throw throw x()`

### second example even simpler
open your console and write `throw notExistingVar` : your code is exactly the same as `notExistingVar` because your code fails at runtime **before** running the `throw` because the right operand is in error, still writing `throw notExistingVar` is allowed

@rbuckton @ljharb @Jack-Works : Is it clear enough to close this issue ?

**Comment by Jack-Works** [comment]
Yes, I don't think it should be a language problem. Maybe eslint can ban that

**Comment by ljharb** [comment]
eslint absolutely would :-) since the committee expressed what i'd describe as a strong unwillingness to try to ban these sorts of useless expressions (due to the grammar complexities involved, to be fair), it seems that it will be left to linters and styleguides to ban them.

**Comment by MadaraUchiha** [comment]
Agreed. We already have precedence in the form of empty expressions and so on which is valid JavaScript but disallowed by linters.

**Comment by rbuckton** [comment]
I concur with the general consensus that there is no reason to ban `throw throw` in the language itself and to leave it to linters/etc. to enforce any restrictions.
